File: OmarPakker/Source/QL_GOLD_C-Sharp/Algebra.Core/Expr/BinaryExpr.cs
using Algebra.Core.Tree;

namespace Algebra.Core.Expr
{
    public abstract class BinaryExpr<E> : UnaryExpr<E>
        where E : IExprNode
    {
        public E Expr2 { get; private set; }

        public BinaryExpr(E e1, E e2)
            : base(e1)
        {
            Expr2 = e2;
        }

		public override bool Equals(object obj)
		{
			if (!base.Equals(obj) || !(obj is BinaryExpr<E>))
			{
				return false;
			}

			return Expr2.Equals(((BinaryExpr<E>)obj).Expr2);
		}

		public override int GetHashCode()
		{
			return base.GetHashCode() + Expr2.GetHashCode();
		}
    }
}

File: GeorgePachitariu/ANTLR-First/src/expr/BinaryExpr.java
package expr;

public abstract class BinaryExpr extends Expression {
	protected Expression leftHandOperand;
	protected Expression rightHandOperand;


	public BinaryExpr(Expression first, Expression second) {
		this.leftHandOperand=first;
		this.rightHandOperand=second;
	}

	@Override
	public boolean equals(Object obj) {
		if (!(obj instanceof BinaryExpr))
			return false;
		BinaryExpr binaryExpr=(BinaryExpr) obj;
		return this.leftHandOperand.equals(binaryExpr.leftHandOperand) &&
				this.rightHandOperand.equals(binaryExpr.rightHandOperand);
	}	

	@Override
	public boolean containsTreeNode(Expression e) {
		if(e.equals(this) || 
				this.leftHandOperand.containsTreeNode(e) ||
				this.rightHandOperand.containsTreeNode(e) )
			return true;
		return false;
	}
}
